### ISSUE TITLE
async actix-web clients fail to build

### ISSUE THREAD

**Issue by ensc** [ramosbugs/oauth2-rs#80]
When trying to build an actix-web based client for the http requiest, I get errors about missing `Send` markers. E.g. a stripped down example might be

```rust
async fn request()
{
    let mut response = awc::ClientBuilder::new()
	.finish()
	.get("")
	.send_body("")
	.await
	.unwrap();

    response.body().await.unwrap();
}

async fn request_async<FN, F>(f: FN) -> ()
where
    FN: FnOnce() -> F + Send,
    F:  std::future::Future<Output = ()> + Send,  /// <<<< removing this Send makes it work
{
    f().await
}

async fn auth()
{
    request_async(request).await;
}

fn main() {
}
```

which results into

```
error[E0277]: `std::rc::Rc<awc::ClientConfig>` cannot be sent between threads safely
   = help: within `impl std::future::Future`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<awc::ClientConfig>`
```

(and much more).

The `request_auth()` function corresponds to oauth2:'s methods which have a similar signature.

When removing the `Send` requirement from `F`, things seem to be fine.  In oauth2 this seems to be required due to `#[async_trait]` but accordingly its [documentation](https://docs.rs/async-trait/0.1.22/async_trait/#non-threadsafe-futures) this can be explicitly disabled by writing

```rust
#[async_trait(?Send)]
```

Is there a reason why `Send` is required for the `http_client` or would it be possible to avoid it (which makes writing awc clients much easier)?

rust-1.40.0
awc-1.0.1
actix-web-2.0 (futures-0.3)
oauth2-3.0.0-alpha.7

**Comment by ramosbugs** [comment]
Hmm where does the `Rc` come from? wouldn't making it an `Arc` solve the problem?

I don't fully understand what a "non-threadsafe future" would be... that seems somewhat contradictory, no?

**Comment by ensc** [comment]
The `Rc` is coming from actix and there are yet more more objects which are not thread safe.  There is probably more work required than replacing `Rc` by `Arc` to change this.

I have no idea whether a non-threadsafe future would be contradictory. The `async` functions (the `rust` language feature) do not have the `Send` requirement; it was explicitly added by the `async_trait` crate (with a way to opt-out).  

Perhaps the executor sticks such futures to a thread while `Send` futures can be assigned to other ones.

**Comment by asonix** [comment]
The Axtix runtime is made up of multiple threads each running the tokio current-thread runtime. Since threads in Actix don't share work, it is fine for futures to contain !Send types.

**Comment by ramosbugs** [comment]
after chatting with a coworker, it sounds like we can't safely remove the `Send` bound without breaking compatibility with executors that require `Send` `Future`s. Apparently, the return value of the `async` trait function gets desugared to `Box<dyn Future<Result<TR, RequestTokenError<RE, TE>>> + Send>`. Because of the type erasure, if we remove `Send` from this type, then the compiler won't know the future is `Send` even if it is. That would preclude using any multi-threaded executors (which require `Send`).

So, we can't unconditionally remove the `Send` bound, but we could make it a feature flag or something similar. I would be open to a PR that does that as long as it doesn't duplicate too much code or make things too messy.

**Comment by fakeshadow** [comment]
I encounter this situation where `Send` bound get in the way when I want specific `!Send` futures.
I imagine we can transform the `async_internal` module into a gaint macro and change the async_trait attribute along with the `Send` bound according to an additional feature flag. 
It would add like a couple loc and an optional feature.The downside I can think of is the macro would give IDEs problem for hinting and auto completing for said module. 

I can make a PR if this is acceptable.

**Comment by mettke** [comment]
@ramosbugs Unfortunately I have to somewhat disagree with your coworkers. They are right when it comes to using async-trait. We cannot remove the Send flag as it will not work then. But we put this onto ourselves by using async-trait in the first place (even thou it should not be necessary at all).

Before I say any more words let me show you two examples:

```
use std::rc::Rc;
use std::time::Duration;

use tokio::time::delay_for;

async fn request() -> () {
    let rc = Rc::new(0);
    println!("Before sleep");
    delay_for(Duration::from_millis(500)).await;
    println!("After sleep");
}

async fn request_async<FN, F>(f: FN) -> ()
where
    FN: FnOnce() -> F,
    F: std::future::Future<Output = ()>,
{
    f().await
}

#[tokio::main]
async fn main() -> Result<(), std::io::Error> {
    println!("Start");
    tokio::spawn(request_async(request));
    println!("End");
    Ok(())
}
```
This one cannot be compiled as tokio requires futures which are Send and the future request returns is not Send (Rc). This one however does compile:
```
use std::sync::Arc;
use std::time::Duration;

use tokio::time::delay_for;

async fn request() -> () {
    let rc = Arc::new(0);
    println!("Before sleep");
    delay_for(Duration::from_millis(500)).await;
    println!("After sleep");
}

async fn request_async<FN, F>(f: FN) -> ()
where
    FN: FnOnce() -> F,
    F: std::future::Future<Output = ()>,
{
    f().await
}

#[tokio::main]
async fn main() -> Result<(), std::io::Error> {
    println!("Start");
    tokio::spawn(request_async(request));
    println!("End");
    Ok(())
}
```
As request does return a Sync Future. Do you notice something? Both do not require Send. The first is rejected for not having Send and the second one is accepted for having Send. Why that? Because send is automatically implemented for any type which consist out of Send Types.

Ok fine, but why does it not work with async-trait? Easy, async in traits is not fully implemented in rust yet and async-trait is a hacky workaround to make it somewhat work. But its limited to the current functionality of the language as thus is imperfect and we cannot expect the same promise that the language normally gives us. 

Long story short, what can we do:
Refactor our async code to remove the async_trait. Of course this takes away the beauty of being able to implement a trait to provide a given functionality but I guess that it is ok as we do not expect that trait anywhere. It makes using this library a lot easier.

**Comment by fakeshadow** [comment]
@mettke This looks good. It would possibly make the code faster too since trait method have to box the future as return type which lead to more heap allocation.

**Comment by ramosbugs** [comment]
The `async_internal` module and its use of `async_trait` was originally a hack to continue supporting old, pre-async-await versions of rustc, which wouldn't compile `async` methods even if gated behind `#[cfg(...)]`. Putting that code into a separate module and gating the entire module behind `#[cfg(...)]` did compile, however.

Since the 4.x release of this crate is dropping support for those rust versions, I _think_ we can get rid of the traits now and make those functions be `impl` methods of `CodeTokenRequest`, etc. If I understand the discussion above, dropping `async_trait` would remove the type erasure (`Box<dyn Future<...>>`) that was causing `Send` to get dropped if it's not explicitly part of the bound on the callback type. So, I think that'll make everyone happy?

**Comment by mettke** [comment]
I created a PR which should solve the problem for the `actix-web` client and still works in threaded async runtimes
